nucliadb/nucliadb/ingest/orm/__init__.py

```diff
@@ -45,38 +44,44 @@ def __init__(self):
     def get_local_node(self):
         return self.local_node
 
-    def find_nodes(self, kbid: str):
-        if self.cluster is not None:
-            nodes = self.cluster.find_nodes(kbid)
-            if len(nodes) != len(set(nodes)):
-                raise NodeClusterSmall()
-            if len(nodes) < settings.node_replicas:
-                raise NodeClusterSmall()
+    def find_nodes(self, exclude_nodes: Optional[List[str]] = None) -> List[str]:
+        """
+        Returns a list of node ids sorted by increasing shard count and load score.
+        It will exclude the node ids in `excluded_nodes` from the computation.
+        It raises an exception if it can't find enough nodes for the configured replicas.
+        """
+        target_replicas = settings.node_replicas
+        total_nodes = len(NODES)
+        if total_nodes < target_replicas:
+            raise NodeClusterSmall(
+                f"Not enough nodes. Total: {total_nodes}, Required: {target_replicas}"
+            )
 
-            return nodes[: settings.node_replicas]
-        else:
-            raise NodeClusterNotFound()
+        available_nodes = [
```

Comment:
NIT: If all filtering conditions are met, we'll iterate the list of nodes 4 times + 1 sorting. Might be good to iterate less times and filters easier to understand (filtering by x[0] or x[1] is more difficult to read than node_id or node.shard_count)

Reply:
The main reason to iterate the nodes multiple times was to be able to provide a different error message depending on the reason for not finding enough nodes.

I agree with the filtering being not too readable though. will change that